src/object/obj_tx.c

```diff
@@ -792,154 +841,815 @@ dc_tx_open(tse_task_t *task)
 }
 
 struct tx_commit_cb_args {
-	struct dc_tx	*tcca_tx;
-	crt_rpc_t	*tcca_req;
+	struct dc_tx		*tcca_tx;
+	crt_rpc_t		*tcca_req;
+	daos_tx_commit_t	*tcca_args;
 };
 
 static int
 dc_tx_commit_cb(tse_task_t *task, void *data)
 {
-	struct tx_commit_cb_args *tcca = (struct tx_commit_cb_args *)data;
+	struct tx_commit_cb_args *tcca = data;
 	struct dc_tx		 *tx = tcca->tcca_tx;
 	crt_rpc_t		 *req = tcca->tcca_req;
+	struct obj_cpd_out	 *oco = crt_reply_get(req);
+	tse_task_t		 *pool_task = NULL;
+	uint64_t		 *sub_epochs;
 	int			  rc = task->dt_result;
+	int			  rc1;
+	bool			  locked = true;
 
 	D_MUTEX_LOCK(&tx->tx_lock);
 
+	if (rc == 0) {
+		rc = oco->oco_ret;
+		if (rc == 0) {
+			int	*sub_rets = oco->oco_sub_rets.ca_arrays;
+
+			/* FIXME: currently, we pack one DTX per CPD RPC. */
+			rc = sub_rets[0];
+		}
+	}
```

Comment:
If the CPD RPC failed for some reason, the oco_map_version may be not set, under such case, it will be zero (the initial value). Then the check "if (tx->tx_pm_ver < oco->oco_map_version)" will be false.

If the CPD RPC return-value is neither "-DER_TX_RESTART" nor "-DER_STALE", then we will resend the CPD RPC if it is obj_retry_error().

For "-DER_STALE" case, we do not know whether some daos server(s) that will be touched by current DTX have been changed because of the pool map refresh, so we have to ask the DTX to restart.

Reply:
> If the CPD RPC failed for some reason, the oco_map_version may be not set, under such case, it will be zero (the initial value). Then the check "if (tx->tx_pm_ver < oco->oco_map_version)" will be false.

If the RPC didn't get a reply, not reading even the initial value in the reply buffer would be cleaner. I don't insist a change though.

> If the CPD RPC return-value is neither "-DER_TX_RESTART" nor "-DER_STALE", then we will resend the CPD RPC if it is obj_retry_error().

OK, I see that now.

> For "-DER_STALE" case, we do not know whether some daos server(s) that will be touched by current DTX have been changed because of the pool map refresh, so we have to ask the DTX to restart.

OK.